### PR TITLE
Unit testing

### PR DESCRIPTION
**Pull request by IsaiahV98**

### CODE REVIEW

**Comment by IsaiahV98** [comment]
## Focus: NestJS/TypeORM Unit Testing 
Initially I utilized the following guide: [“_Unit Testing with NestJS and Jest: A comprehensive tutorial_” by jackallcock97](https://medium.com/@jackallcock97/unit-testing-with-nestjs-and-jest-a-comprehensive-tutorial-464910f6c6ba).
### Highlights: ###
- Test a Controller by mocking the services via  jest.mock("../.../foobar.service"); after each import. 
`jest.mock("../client/client.service");`
- Test a service by mocking the repositories via the use of custom mocks. 
```// @ts-ignore 
export const repositoryMockFactory: () => MockType<Repository<any>> = jest.fn(() => ({
 findOne: jest.fn(),
 find: jest.fn(),
 update: jest.fn(),
 save: jest.fn()
}));export type MockType<T> = {
 [P in keyof T]: jest.Mock<{}>;
};
```

### Concerns: ### 
- Introduction of clutter via custom mocks.

- High possibility of accidentally introducing [Implementation testing](https://kentcdodds.com/blog/testing-implementation-details).

These concerns lead me to seeking out other resource for information, which led me to find the work of Jay McDoniel on [Stack Overflow](https://stackoverflow.com/questions/57099863/spyon-typeorm-repository-to-change-the-return-value-for-unit-testing-nestjs
) and [GitHub](https://github.com/jmcdo29/testing-nestjs).

### Highlights: ####
- Cleaner implementation of Unit Testing for Services without the need of custom mocks. 

- Provided detailed examples of both strictly Unit Testing, and Integration Testing. 


While there still remains a decent amount of logic without coverage in the Covid Data-Api folder, the plan is to cover these functions in the future with [Integration Testing](https://github.com/rostgoat/testing-nestjs/blob/9dbc1a4810ce0ecffc2399eaaf2b7cdee03b89db/apps/typeorm-sample/src/cat/cat.integration.spec.ts
) to avoid writing Unit Tests that are implementation dependent. 


### Other Notes: ###  
I have chosen to wrap every instance of testing a specific logical function call in a **Describe** statement, with **It** statements for every logical return case inside. This allows for clearer reading of both the file, and the output from Jest. 
```
  describe('getWebsitesForCounty', () => {
    it('should return expectedResult', async () => {
      const mockParameter = 'undefined';
      const expectedResult = {
        status: 400,
        success: false,
        message: 'Invalid county fips'
      };
      expect(await websitesService.getWebsitesForCounty(mockParameter)).toMatchObject(expectedResult);
    });
    it('should return expectedResult', async () => {
      const mockParameter = undefined;
      const expectedResult = {
        status: 400,
        success: false,
        message: 'Invalid county fips'
      };
      expect(await websitesService.getWebsitesForCounty(mockParameter)).toMatchObject(expectedResult);
    });
    it('should return expectedResult', async () => {
      const mockParameter = 0;
      const expectedResult = {
        status: 400,
        success: false,
        message: 'Invalid county fips'
      };
      expect(await websitesService.getWebsitesForCounty(mockParameter)).toMatchObject(expectedResult);
    });
  });
```

**Review by Proantagonist — CHANGES_REQUESTED** [review]
Overall solid PR.

Only minor changes in this one. 

For future PR's, or passes keep in mind that form components (input, texbox, etc) need to test interoperability with form controls for use in reactive forms. This is their main intended use case.

In terms of style, keep it consistent and let things breathe. Makes it easier to read. 

```
  it('should correctly evaluate registerOnChange', () => {
    component.registerOnChange(component.writeValue(true));
    expect(component.value).toBeTruthy();
  });
  it('should correctly evaluate registerOnTouched', () => {
    component.registerOnTouched(component.writeValue(true));
    expect(component.value).toBeTruthy();
  });
  it('should correctly evaluate setDisabledState', () => {
    component.setDisabledState(true);
    expect(component.disabled).toBeTruthy();
  });
```

Adding a space between each of those cases increases readability.

File: libs/ui-kits/ngx/layout/src/lib/components/accordion/accordion.component.spec.ts [hunk 1]
```diff
@@ -1,13 +1,11 @@
...
 import { AccordionComponent } from './accordion.component';
-import { Component, ViewChild } from '@angular/core';
-import { AccordionHeaderComponent } from './accordion-header/accordion-header.component';
-import { AccordionContentComponent } from './accordion-content/accordion-content.component';
+
 import { AccordionService } from './services/accordion.service';
 
 // Mock testing component to emulate ContentChildren title and content components.
-@Component({
+/*@Component({
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/layout/src/lib/components/accordion/accordion.component.spec.ts:8]: Removed

File: libs/covid/data-api/src/lib/modules/base/base.service.spec.ts [hunk 2]
```diff
@@ -1,21 +1,35 @@
...
   });
+  /* Implementation Testing?
+  describe('getOne ', () => {
+    it('Service should be defined', async () => {
+      const mockedParamater = {};
+      const expectedResult = new BaseEntity();
+      jest.spyOn(baseService.repository, 'findOne').mockResolvedValue(expectedResult);
+      expect(await baseService.getOne(mockedParamater)).toEqual(expectedResult);
+    });
+  });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/base/base.service.spec.ts:34]: Removed

File: libs/covid/data-api/src/lib/modules/counties/counties.service.spec.ts [hunk 3]
```diff
@@ -19,10 +21,70 @@ describe('CountiesService', () => {
...
+      hasId: undefined,
+      save: undefined,
+      remove: undefined,
+      reload: undefined
+    };
+    const countyTest = new County();
+    countyTest.stateFips = stateFipsMock;
+    CountyMockRepo.find.mockReturnValue(countyTest);
+    expect(service.getCountiesForState('yeet')).toBe(countyTest);
+  });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/counties/counties.service.spec.ts:89]: Removed

File: libs/covid/data-api/src/lib/modules/county-claims/county-claims.service.spec.ts [hunk 4]
```diff
@@ -22,11 +28,92 @@ describe('CountyClaimsService', () => {
...
+    it('should handle mockParameter being defined, but spyOn undefined ', async () => {
+      const mockParameter = 'foobar';
+      const expectedResult = {
+        status: 500,
+        success: false,
+        message: 'Invalid claim.'
+      };
+      jest.spyOn(countyClaimRepo, 'findOne').mockReturnValue(undefined);
+      expect(await countyClaimsService.closeClaim(mockParameter)).toMatchObject(expectedResult);
+    });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/county-claims/county-claims.service.spec.ts:93]: Removed

File: libs/covid/data-api/src/lib/modules/field-categories/field-categories.service.spec.ts [hunk 5]
```diff
@@ -20,10 +22,105 @@ describe('FieldCategoriesService', () => {
...
+    });
+
+    it('should handle catagorey inputs ', async () => {
+      FieldCategoryMockRepo.findOne.mockReturnValue(undefined);
+      expect(await service.getFieldTypesForCategory(9)).toMatchObject({
+        status: 400,
+        success: false,
+        message: 'Invalid category ID.'
+      });
+    });
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/field-categories/field-categories.service.spec.ts:72]: Removed

File: libs/covid/data-api/src/lib/modules/field-categories/field-categories.service.spec.ts [hunk 6]
```diff
@@ -20,10 +22,105 @@ describe('FieldCategoriesService', () => {
...
+    /*it('should handle catagorey inputs ', async () => {
+      let categ = new CategoryValue();
+      categ.category = new FieldCategory();
+      categ.category.types = [new FieldType()];
+
+      FieldCategoryMockRepo.findOne.mockReturnValueOnce(categ);
+      FieldTypeMockRepo.findOne.mockReturnValue(categ);
+
+      expect(await service.addFieldTypeToCategory(9, 'yeet')).toMatchObject(categ);
+    });
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/field-categories/field-categories.service.spec.ts:98]: Removed

File: libs/covid/data-api/src/lib/modules/field-categories/field-categories.service.spec.ts [hunk 7]
```diff
@@ -20,10 +22,105 @@ describe('FieldCategoriesService', () => {
...
+    it('should handle catagorey inputs ', async () => {
+      let categ = new CategoryValue();
+      categ.category = new FieldCategory();
+      categ.category.types = [new FieldType()];
+
+      FieldCategoryMockRepo.findOne.mockReturnValueOnce(categ);
+      FieldTypeMockRepo.findOne.mockReturnValue(categ);
+
+      expect(await service.addFieldTypeToCategory(9, 'yeet')).toMatchObject(categ);
+    });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/field-categories/field-categories.service.spec.ts:125]: Removed

File: libs/covid/data-api/src/lib/modules/phone-numbers/phone-numbers.service.spec.ts [hunk 8]
```diff
@@ -22,10 +27,64 @@ describe('PhoneNumbersService', () => {
...
+      };
+      expect(await phoneNumbersService.getPhoneNumbersForCounty(mockParameter)).toMatchObject(expectedResult);
+    });
+    /* Implementation Testing?
+    it('should return expectedResult ', async () => {
+      const mockParameter = 'yooo';
+      const expectedResult = 'foobar';
+      jest.spyOn(phoneNumbersService as any, 'getPhoneNumbers').mockResolvedValue(expectedResult);
+      expect(await phoneNumbersService.getPhoneNumbersForCounty(mockParameter)).toEqual(expectedResult);
+    });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/phone-numbers/phone-numbers.service.spec.ts:76]: Removed

File: libs/covid/data-api/src/lib/modules/states/states.service.spec.ts [hunk 9]
```diff
@@ -8,16 +8,39 @@ import { State } from '@tamu-gisc/covid/common/entities';
...
+    });*/
+  });
+  describe('getStateByFips', () => {
+    /* Implementation Testing?
+    it('should getStateByFips', async () => {
+      const mockParameter = 99;
+      const expectedResult = new State();
+      jest.spyOn(stateMockRepo, 'findOne').mockResolvedValue(expectedResult);
+      expect(await statesService.getStateByFips(mockParameter)).toBe(expectedResult);
+    });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/states/states.service.spec.ts:44]: Removed

[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/states/states.service.spec.ts:44]: Removed

File: libs/covid/data-api/src/lib/modules/testing-sites/testing-sites.service.spec.ts [hunk 10]
```diff
@@ -37,11 +48,57 @@ describe('TestingSitesService', () => {
...
+    });
+  });
+  describe('getWebsitesForClaimInfo', () => {
+    /* Implementation Testing?
+    it('should throw error with mockParameter being Empty String ', async () => {
+      const mockParameter = '';
+      jest.spyOn(userMockRepo, 'findOne').mockResolvedValue(undefined);
+      await expect(testingSitesService.createOrUpdateTestingSite(mockParameter)).rejects.toThrow();
+    });*/
+  });
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/testing-sites/testing-sites.service.spec.ts:76]: Removed

File: libs/covid/data-api/src/lib/modules/users/users.service.spec.ts [hunk 11]
```diff
@@ -19,10 +21,80 @@ describe('UsersService', () => {
...
+
+  /* Implementation Testing?
+  describe('getUsers', () => {
+    it('should return expectedResult', async () => {
+      const expectedResult = [];
+      jest.spyOn(userMockRepo, 'find').mockResolvedValue(expectedResult);
+      expect(await usersService.getUsers()).toEqual(expectedResult);
+    });
+  });
+  */
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/users/users.service.spec.ts:43]: Removed

File: libs/covid/data-api/src/lib/modules/users/users.service.spec.ts [hunk 12]
```diff
@@ -19,10 +21,80 @@ describe('UsersService', () => {
...
+      expect(await usersService.verifyEmail(mockParameter)).toMatchObject(expectedResult);
+    });*/
+
+    /* Implementation Testing?
+    it('should return user if findOne returns, and if email length > 0 ', async () => {
+      const mockParameter = 'Foobar';
+      const expectedResult = new User();
+      jest.spyOn(userMockRepo, 'findOne').mockResolvedValue(expectedResult);
+      expect(await usersService.verifyEmail(mockParameter)).toEqual(expectedResult);
+    });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/users/users.service.spec.ts:76]: Removed

File: libs/covid/data-api/src/lib/modules/users/users.service.spec.ts [hunk 13]
```diff
@@ -19,10 +21,80 @@ describe('UsersService', () => {
...
+      expect(await usersService.registerEmail(mockParameterUndefined)).toMatchObject(expectedResult);
+    });
+
+    /* Implementation Testing?
+    it('should return user if findOne returns, and if email length > 0 ', async () => {
+      const mockParameter = 'foobar';
+      const expectedResult = new User();
+      jest.spyOn(userMockRepo, 'findOne').mockResolvedValue(expectedResult);
+      expect(await usersService.registerEmail(mockParameter)).toEqual(expectedResult);
+    });*/
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/covid/data-api/src/lib/modules/users/users.service.spec.ts:98]: Removed

File: libs/ui-kits/ngx/forms/src/lib/components/checkbox-group/checkbox-group.component.spec.ts [hunk 14]
```diff
@@ -1,6 +1,7 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
 import { CheckboxGroupComponent } from './checkbox-group.component';
+import { QueryList } from '@angular/core';
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/forms/src/lib/components/checkbox-group/checkbox-group.component.spec.ts:4]: Removed unused imports

File: libs/ui-kits/ngx/forms/src/lib/components/radio-group/radio-group.component.spec.ts [hunk 15]
```diff
@@ -1,5 +1,7 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
+import { Angulartics2Module } from 'angulartics2';
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/forms/src/lib/components/radio-group/radio-group.component.spec.ts:3]: Removed unused imports

File: libs/ui-kits/ngx/layout/src/lib/components/accordion/accordion-header/accordion-header.component.spec.ts [hunk 16]
```diff
@@ -2,21 +2,35 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
 import { AccordionHeaderComponent } from './accordion-header.component';
 import { AccordionService } from '../services/accordion.service';
+import { BehaviorSubject, Observable, of as observableOf } from 'rxjs';
+import { EventEmitter } from 'events';
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/layout/src/lib/components/accordion/accordion-header/accordion-header.component.spec.ts:6]: Removed unused imports, and auto-imports

File: libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile-navigation/tile-navigation.component.spec.ts [hunk 17]
```diff
@@ -3,25 +3,57 @@ import { RouterTestingModule } from '@angular/router/testing';
 
 import { TileNavigationComponent } from './tile-navigation.component';
 import { TileSubmenuContainerComponent } from '../tile-submenu-container/tile-submenu-container.component';
+import { Observable, of as observableOf } from 'rxjs';
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile-navigation/tile-navigation.component.spec.ts:6]: Remove unused imports.

[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile-navigation/tile-navigation.component.spec.ts:6]: Import statements correctly ordered

File: libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile-submenu/tile-submenu.component.spec.ts [hunk 18]
```diff
@@ -1,24 +1,43 @@
...
+    TileServiceStub = TestBed.get(TileService);
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should create', () => {
+    const yeet = spyOn(TileServiceStub, 'toggleSubmenu');
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile-submenu/tile-submenu.component.spec.ts:39]: Fixed

File: libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile/tile.component.spec.ts [hunk 19]
```diff
@@ -1,24 +1,42 @@
...
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/navigation/mobile-tile/src/lib/components/tile/tile.component.spec.ts:41]: Deleted

File: libs/ui-kits/ngx/navigation/mobile-tile/src/lib/services/tile.service.spec.ts [hunk 20]
```diff
@@ -1,12 +1,41 @@
...
+    service.toggleSubmenu();
+    expect(service.submenuActive.value).toBeFalsy();
+  });
+
+  it('should toggleSubmenu', () => {
+    service.toggleSubmenu(false);
+    expect(service.submenuActive.value).toBeFalsy();
+    service.toggleMenu(false);
+    expect(service.submenuActive.value).toBeFalsy();
+  });
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/navigation/mobile-tile/src/lib/services/tile.service.spec.ts:34]: Described behavior.

File: libs/ui-kits/ngx/layout/src/lib/abstracts/abstract-content-swap/abstract-content-replacer.component.spec.ts [hunk 21]
```diff
@@ -22,4 +21,13 @@ describe('AbstractContentReplacerComponent', () => {
...
     expect(component).toBeTruthy();
   });
+
+  it('should use SwapContent correctly', () => {
+    component.swapContent({ index: 9, label: 'yeet', template: undefined });
+    component.swapContent({ index: 9, label: 'yeet', template: undefined });
+    expect(component.toggleIndex).toEqual(9);
+    component.swapContent(8);
+    component.swapContent(8);
+    expect(component.toggleIndex).toEqual(8);
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/layout/src/lib/abstracts/abstract-content-swap/abstract-content-replacer.component.spec.ts:31]: I believe this was something that I was meaning to get back to, but overlooked. Fixed now.

File: libs/ui-kits/ngx/layout/src/lib/components/accordion/services/accordion.service.spec.ts [hunk 22]
```diff
@@ -3,14 +3,57 @@ import { TestBed } from '@angular/core/testing';
...
+        _isScalar: false,
+        _value: { animate: true, expanded: true, resize: true },
+        closed: false,
+        hasError: false,
+        isStopped: false,
+        observers: [],
+        thrownError: null
+      }
+    });
+  });
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/layout/src/lib/components/accordion/services/accordion.service.spec.ts]: Fixed

[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/ui-kits/ngx/layout/src/lib/components/accordion/services/accordion.service.spec.ts]: Checked emitted values

[the remark this answers is not in the document]
Comment (Proantagonist) [reply libs/ui-kits/ngx/layout/src/lib/components/accordion/services/accordion.service.spec.ts]: 👍

**Pull request closed without merging**